cmd/metal-api/internal/service/project-service.go

```diff
@@ -161,34 +174,221 @@ func (r projectResource) findProjects(request *restful.Request, response *restfu
+		return
+	}
+	if len(ns) > 0 {
+		checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("there are still networks allocated by this project"))
+		return
+	}
+
+	var ips metal.IPs
+	err = r.ds.SearchIPs(&datastore.IPSearchQuery{ProjectID: &id}, &ips)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	if len(ips) > 0 {
+		checkError(request, response, utils.CurrentFuncName(), fmt.Errorf("there are still ips allocated by this project"))
+		return
+	}
+
+	pdr := &mdmv1.ProjectDeleteRequest{
+		Id: p.Project.Meta.Id,
+	}
+	pdresponse, err := r.mdc.Project().Delete(context.Background(), pdr)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	v1p := mapper.ToV1Project(pdresponse.Project)
+	pcres := &v1.ProjectResponse{
+		Project: *v1p,
+	}
+
+	err = response.WriteHeaderAndEntity(http.StatusOK, pcres)
+	if err != nil {
+		zapup.MustRootLogger().Error("Failed to send response", zap.Error(err))
+		return
+	}
+}
+
+func (r projectResource) updateProject(request *restful.Request, response *restful.Response) {
+	var requestPayload v1.ProjectUpdateRequest
+	err := request.ReadEntity(&requestPayload)
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+	if requestPayload.Project.Meta == nil {
+		err = errors.New("project and project.meta must be specified")
+		checkError(request, response, utils.CurrentFuncName(), err)
+		return
+	}
+
+	existingProject, err := r.mdc.Project().Get(context.Background(), &mdmv1.ProjectGetRequest{Id: requestPayload.Project.Meta.Id})
+	if checkError(request, response, utils.CurrentFuncName(), err) {
+		return
+	}
+
+	// new data
+	projectUpdateData := mapper.ToMdmV1Project(&requestPayload.Project)
+	// created date is not updateable
+	projectUpdateData.Meta.CreatedTime = existingProject.Project.Meta.CreatedTime
+
+	pur, err := r.mdc.Project().Update(context.TODO(), &mdmv1.ProjectUpdateRequest{
+		Project: projectUpdateData,
+	})
+	if err != nil {
+		checkError(request, response, utils.CurrentFuncName(), err)
+		return
+	}
+
+	v1p := mapper.ToV1Project(pur.Project)
+	err = response.WriteHeaderAndEntity(http.StatusOK, v1.ProjectResponse{
+		Project: *v1p,
+	})
+	if err != nil {
+		zapup.MustRootLogger().Error("Failed to send response", zap.Error(err))
+		return
+	}
+}
+
+func (r projectResource) ipsByProject() (map[string]metal.IPs, error) {
+	ips, err := r.ds.ListIPs()
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]metal.IPs)
+	for _, ip := range ips {
+		if ip.ProjectID == "" {
+			continue
+		}
+		pips := result[ip.ProjectID]
+		pips = append(pips, ip)
+		result[ip.ProjectID] = pips
+	}
+	return result, nil
+}
+
+func (r projectResource) machinesByProject() (map[string]metal.Machines, error) {
+	// FIXME implement ScopedMachineFind
+	ms, err := r.ds.ListMachines()
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]metal.Machines)
```

Comment:
There actually is `ByProjectID()` on the `Machines` struct, which is the result of `SearchMachines`

Reply:
done

Reply:
use ByProjectID() now